backends-velox/src/main/scala/org/apache/gluten/execution/ColumnarPartialGenerateExec.scala

```diff
@@ -0,0 +1,383 @@
+import org.apache.spark.sql.execution.{ExplainUtils, GenerateExec, SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.sql.types.{ArrayType, BinaryType, DataType, MapType, StringType, StructType}
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+import scala.collection.mutable.{ArrayBuffer, ListBuffer}
+
+/**
+ * By rule <PartialGenerateRule>, if the generator is a instance of <HiveGenericUDTF>, then the
+ * generateExec will be changed to ColumnarPartialGenerateExec
+ *
+ * @param generateExec
+ *   the GenerateExec from vanilla
+ * @param child
+ *   child plan
+ */
+case class ColumnarPartialGenerateExec(generateExec: GenerateExec, child: SparkPlan)
+  extends UnaryExecNode
+  with ValidatablePlan {
+
+  private val generatorNullRow = new GenericInternalRow(generateExec.generatorOutput.length)
+
+  private val pruneChildAttributes: ListBuffer[Attribute] = ListBuffer()
+  private val pruneChildColumnIndices: ListBuffer[Int] = ListBuffer()
+  private val generateAttributes: ListBuffer[Attribute] = ListBuffer()
+  private val rightInputColumnIndices: ListBuffer[Int] = ListBuffer()
+
+  private val rightSchema =
+    SparkShimLoader.getSparkShims.structFromAttributes(generateExec.generatorOutput)
+
+  private lazy val getLeftIndices = getColumnIndexInChildOutput(
+    pruneChildAttributes,
+    pruneChildColumnIndices,
+    generateExec.requiredChildOutput)
+  private lazy val getRightIndices = getColumnIndexInChildOutput(
+    generateAttributes,
+    rightInputColumnIndices,
+    Seq(generateExec.generator))
+
+  private lazy val generator = InterpretedArrowGenerate.create(
+    bindReferences(Seq(generateExec.generator), generateAttributes.toSeq).head)
+
+  @transient override lazy val metrics = Map(
+    "time" -> SQLMetrics.createTimingMetric(sparkContext, "total time of partial project"),
+    "velox_to_arrow_time" -> SQLMetrics.createTimingMetric(
+      sparkContext,
+      "time of velox to Arrow ColumnarBatch"),
+    "arrow_to_velox_time" -> SQLMetrics.createTimingMetric(
+      sparkContext,
+      "time of Arrow ColumnarBatch to velox")
+  )
+
+  private def getColumnIndexInChildOutput(
+      attributes: ListBuffer[Attribute],
+      indices: ListBuffer[Int],
+      exprs: Seq[Expression]): Unit = {
+    exprs.forall {
+      case a: AttributeReference =>
+        val index = child.output.indexWhere(s => s.exprId.equals(a.exprId))
+
+        if (index < 0) {
+          throw new IllegalStateException(
+            s"Couldn't find $a in ${child.output.attrs.mkString("[", ",", "]")}")
+        } else if (!indices.contains(index)) {
+          attributes.append(a)
+          indices.append(index)
+          true
+        } else true
+      case p =>
+        getColumnIndexInChildOutput(attributes, indices, p.children)
+        true
+    }
+  }
+
+  override protected def doValidateInternal(): ValidationResult = {
+    ValidationResult.succeeded
```

Comment:
Please respect attrNotExists and hasUnsupportedDataType as PartialProject, some Spark data type is not supported in Velox